In [1]:
from IPython.display import HTML
HTML('''
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.0.3/jquery.min.
js"></script>
<script>
code_show=true;
function code_toggle() {
if (code_show){
$('div.jp-CodeCell > div.jp-Cell-inputWrapper').hide();
} else {
$('div.jp-CodeCell > div.jp-Cell-inputWrapper').show();
}
code_show = !code_show
}
$( document ).ready(code_toggle);
</script>
<form action="javascript:code_toggle()"><input type="submit" 
value="Click here to toggle on/off the raw code."></form>
''')

In [2]:
from IPython.core.display import HTML
HTML("""
<style>
.output_png {
    display: table-cell;
    text-align: center;
    vertical-align: middle;
}
</style>
""")

<h1 style="background-color:#0531BC; color:#F5F5F1; padding: 20px;">Title</h1>

<center><img src="./ML2_FinalProject_Banner.png" width="90%" height="90%"></center>
<center>Image 01. Project Title Banner.</center>

<h1 style="background-color:#0531BC; color:#F5F5F1; padding: 20px;">Executive Summary</h1>

As the pandemic continues to limit jobs and income opportunities, having other sources of income is important. Aside from the usual salaries of the workforce, people have expanded their income streams through other means, such as *investments*. Depending on the risks, investments can grow your money exponentially or empty your wallet. Ideally, investments should aim to **maximize returns** while **minimizing risk**. Even up to the present, portfolio managers have been using the **Mean-Variance Portfolio** (`MVP`) method of portfolio construction to achieve investment objectives. However, through the years, with the advancement of AI, new and novel ways have since been discovered and built by financial engineers. One such well-known strategy is the **Hierarchical Risk Parity** (`HRP`) method of portfolio construction, formally published by *Lopez de Prado* in 2016; the focal point of the strategy is to utilize graph theory and machine learning to allocate portfolio risk properly.

The group's study aims to compare the two portfolio construction methods, MV and HRP, and determine if the HRP portfolio outperforms the MVP method in general market scenarios, albeit limited to a single asset class (US equities). 

For this study, the team took the data of the S&P 500 stocks from Yahoo finance. The data was the daily stock prices from January 2018 to December 2019. After extracting the data, the team performed exploratory data analysis to clean it and present it visually. The team then prepared and constructed the MVP and HRP portfolios from the asset universe. Afterward, the team performed backtesting to compare the results of each portfolio using Sharpe Ratio. The group utilized machine learning to connect the statistical properties of the securities to the difference in Sharpe Ratios between the two strategies. Lastly, the team used post-hoc methods in **explainable AI** (i.e., `SHAP`)  to interpret the differences in portfolio performance. 

Explainable AI (XAI) through Shap allows the group to assess the performance of the two portfolio strategies at a granular level. While the Hierarchical Risk Parity portfolio would not replace the Mean-Variance approach anytime soon, empirical tests prove that the HRP strategy provides a *more diversified portfolio* construction than MVP and not at the expense of returns. In addition, using XAI presents to the investor or asset manager a clearer picture of asset allocation and how each portfolio strategy handles an asset class.

<h1 style="background-color:#0531BC; color:#F5F5F1; padding: 20px;">Highlights</h1>

* The **Hierarchical Risk Parity** Portfolio provides a **more stable diversification strategy** and covers Mean-Variance Portfolio's shortcomings concerning asset allocation.
* Hierarchical Risk Parity Portfolio **performs better on out-of-sample data** than the Mean-Variance approach.
* The use of machine learning and **explainable AI** allows investors and asset managers alike to **examine various strategies at a granular level**.

<h1 style="background-color:#0531BC; color:#F5F5F1; padding: 20px;">Problem Statement</h1>

Can the Hierarchical Risk Parity method replace the Mean-Variance approach as the standard for a "Modern Portfolio"? In what ways can an "asset diversification"-focused approach provide returns which may rival returns-based apprach?

<h1 style="background-color:#0531BC; color:#F5F5F1; padding: 20px;">Motivation</h1>

Investing in stocks, bonds, and other securities has never been as easily accessible. People can now invest and trade through brokers, websites, and even on their phones. In addition, investment is now included in other products like insurance. Technology has opened the market more than ever.

A half-century-old theory is still commonly used in building a portfolio. *Mean-Variance Portfolio (MVP)* is still used widely by the industry because of its simplicity, but with the advancement in Machine Learning and AI, other methods have come up. An example of this new method is the *Hierarchical Risk Parity (HRP) Portfolio*. It is a portfolio-building strategy that utilizes unsupervised learning. Couple with *Explainable AI (XAI)*, people can now analyze the strengths and weaknesses of a portfolio

**Insights to Investments** seeks to take a peak behind a machine learning algorithm and explain how and why one portfolio performs better than the other.

<h1 style="background-color:#0531BC; color:#F5F5F1; padding: 20px;">Introduction</h1>

*Investing* refers to the act of allocating resources (i.e., money) with the expectation of generating a profit in the future. While on the other hand, *portfolio management* is managing a collection of investments to achieve an objective. This objective is to **maximize return** (the potential money you could gain from the investment) and **minimize risk** (the potential money you could lose). A common practice in the industry is that a professional portfolio manager works on behalf of clients. These portfolio managers build a portfolio based on multiple factors and the client's risk appetite. [[1]](https://www.investopedia.com/terms/p/portfoliomanagement.asp)

When talking about portfolio management, these are things that we need to consider: *Asset allocation*, *Diversification*, *Rebalancing*, and *Passive vs. Active Management*. For effective portfolio management, portfolio managers generally distribute the whole investment into several different assets with a specific percentage. To spread the risk across several assets and minimize the effect of unfavorable market conditions. Portfolios have different diversification techniques to achieve it. **Diversification will reduce volatility**, and real diversification happens when investments are allocated across different kinds of securities (i.e., stocks, bonds, and others). *Rebalancing* happens when the portfolio **adjusts its allocations towards its original target allocation** at regular intervals (usually annually). This rebalancing allows to capture of gains and redistributes investment to assets with a high growth potential. [[1]](https://www.investopedia.com/terms/p/portfoliomanagement.asp) Lastly, *Active vs. Passive Management* refers to the strategy of whether the portfolio managers **try to beat the market**. Passive Management, also called *index fund management*, aims to duplicate the results of a specific market, usually indicated by its index. In the Philippines, this index is referred to as the **Philippine Stock Exchange Index (PSEi)**, a collection of 30 listed companies chosen to represent the general movement of the Philippine market. [[2]](https://capital.com/pse-index-psei-definition) On the other hand, Active Management tries to beat the market by strategically buying and selling stocks and other assets. Portfolio managers here must pay more attention to the market trends, shift in the economy, and many other possible indicators of potential changes in the market. [[1]](https://www.investopedia.com/terms/p/portfoliomanagement.asp)

There are many techniques to build a portfolio. For the purpose of the project, we will focus on two: the mean-variance portfolio and the hierarchical risk parity portfolio. **Mean-variance Portfolio** or `MVP` is a part of the modern portfolio theory of American economist Harry Markowitz. In MVP, the risk is measured as the expected return variance. Variance measures how spread out the values of the returns are. It measures the **volatility of the asset in the market**. An investor will typically choose the investment with the higher return given the same variance or a lower variance given the same return. The necessary calculation to build MVP involves optimization to find the efficient frontier. [[3]](https://www.guidedchoice.com/video/dr-harry-markowitz-father-of-modern-portfolio-theory/)

<center><img src="./Efficient_Frontier.jpg" width="60%" height="60%"></center>
<center>Image 02. Mean-Variance Portfolio Efficient Frontier. [3]</center>

With the advancement in Machine Learning, other methods came up, such as the *Hierarchical Risk Parity* (`HRP`) portfolio. It was introduced in a 2016 paper by Marcos López de Prado entitled "Building Diversified Portfolios that Perform Well Out-Of-Sample". It uses an unsupervised learning algorithm to ensure that only similar equities compete for the same portfolio spot. [[4]](https://developer.nvidia.com/blog/hierarchical-risk-parity-on-rapids-an-ml-approach-to-portfolio-allocation/) The algorithm of HRP is broken down into three major steps: *Hierarchical Tree Clustering*, *Matrix Seriation*, and *Recursive Bisection*. A detailed explanation of the algorithm is explained in the linked reference. [[5]](https://developer.nvidia.com/blog/hierarchical-risk-parity-on-rapids-an-ml-approach-to-portfolio-allocation/) What HRP does is that instead of capital allocation where risk is measured to adjust the individual securities, here in HRP risk is in the context of the whole portfolio. HRP wants to ensure a profit regardless of the market condition by diversifying its investment into many industries. More details about the HRP are found in the original paper. [[6]](https://papers.ssrn.com/sol3/papers.cfm?abstract_id=2708678)

The performance of these two portfolios will be compared using the metrics called *Sharpe Ratio*. Proposed by economist William Sharpe, the Sharpe ratio compares the return on an investment and its risk. It assesses the risk and rewards of investment. The exact calculation is given in the image below.

<center><img src="./Sharpe_Ratio.jpg" width="60%" height="60%"></center>
<center>Image 03. Sharpe Ratio Formula. [7]</center>

The numerator is the extra return, the difference between the expected returns and a benchmark (usually the risk-free rate of return). The denominator is the standard deviation of that excess return. [[7]](https://www.investopedia.com/terms/s/sharperatio.asp) The interpretation of its value is how much extra returns you get from the extra risk you take. The higher the value of the Sharpe ratio, the better. A value of *greater than 1* means you are **earning more return than the risk you are taking**, a value of *exactly 1* means you are **earning 1 unit of return for each unit of risk**, and a value of *less than 1* means that you are **earning less return per unit of risk you are taking**.

<h1 style="background-color:#0531BC; color:#F5F5F1; padding: 20px;">Data Source</h1>

***
<h2 style="color:#007CFE">Data Description</h2>

***

Our data source is **Yahoo Finance** (https://finance.yahoo.com/) it delivers credible and reliable daily stock price stats, insights reports, and blog posts along with interviews with leading investment experts. We took the data of the **S&P 500 stocks** for its daily stock prices from *Jan 2018 to Dec 2019*.

<br><center>Table 01. Data Features, Description and Type.</center>
<table style="width:60%">
    <tr style="background-color:#007CFE; color:#F5F5F1">
      <th style="width:30%; text-align:center">Variable Name</th>
      <th style="width:55%; text-align:center" >Data Description</th>
      <th style="width:15%; text-align:center">Data Type</th>
    </tr>
    <tr>
        <td style="text-align:left">Date</td>
        <td style="text-align:left">Date of stock price</td>
        <td style="text-align:left">Datetime</td>
      </tr>
   <tr>
      <td style="text-align:left">Stock Symbol (502 columns)</td>
      <td style="text-align:left">Closing stock price of different companies</td>
      <td style="text-align:left">Float</td>
    </tr>
  </table>

***
<h2 style="color:#007CFE">Data and Project Limitation</h2>

***

The data and project are only limited to equity assets of the `S&P500`, representing the *500 largest companies in the US stock exchange*. Therefore, all portfolio building and analysis will assume that portfolio managers can *only invest in these assets*.

Please also note that the steps followed in this notebook are to replicate the paper of López de Prado that aims to build a diversified portfolio [[6]](https://papers.ssrn.com/sol3/papers.cfm?abstract_id=2708678). Hence, no hyperparameter tuning or more advanced exploration was done in our model, and we focused on interpreting the results to prove the theory.

<h1 style="background-color:#0531BC; color:#F5F5F1; padding: 20px;">Methodology</h1>

<center><img src="./ML2_FinalProject_Methodology.png" width="90%" height="90%"></center>
<center>Image 04. Project Methodology.</center>

**Data Collection and Preprocessing**

We performed *web scraping* to retrieve the daily stock prices of the *S&P 500* at Yahoo Finance and saved it into a CSV file for further analysis and processing.
<br><br>

**Exploratory Data Analysis**

We performed Exploratory Data Analysis (EDA) to understand the raw data if we need to perform further data processing/cleansing, and we've also provided simple and *relevant statistics* about the dataset.
<br><br>

**Train-Test Split**

To prepare for our clustering, as we needed to perform backtesting, we needed to have the train-test split was also where we used the last `25%` of the dataset as our *test set*. We also needed to construct our MVP and HRP datasets from the stocks.
<br><br>

**Portfolio Building**

For MVP, we use the package of `cvxopt` to build the portfolio using optimization. For HRP, we follow the paper.[[6]](https://papers.ssrn.com/sol3/papers.cfm?abstract_id=2708678) First, we need to perform Hierarchical clustering on the prepared dataset. Then,  re-label each value, sort each clustered stock based on distance, calculate the inverse variance portfolio, calculate the variance per cluster, and get the allocation % of each stock and the weights for all types of asset allocations before we can analyze the dataset.
<br><br>

**Portfolio Summary and Comparison**

We performed backtesting with the results of the portfolio building. Then, using the train-test split and *In and Out of Sampling Results*, we plotted the line charts to visualize and easily compare the results between MVP and HRP.
<br><br>

**Data Preparation for Explainable AI**

We need to perform another set of data preparation that would be needed for the `SHAP` analysis, more specifically predicting the `Sharpe ratio`. This involves calculating each stock's mean, standard deviation, drawdown, correlation, and rolling value.
<br><br>

**Explainable AI**

The paper already provided the best model, which is the XGBRegressor. Next, we fit the prepared dataset to forecast the Sharpe ratio. We then use the `TreeExplainer` of `SHAP` to get the top features affecting the portfolio. Finally, we analyzed it using the visualizations of `SHAP` with its summary and dependence plots.

<h1 style="background-color:#0531BC; color:#F5F5F1; padding: 20px;">Data Exploration</h1>

Install `cvxopt` package (if necessary). This will be used for mean-variance portfolio optimization.

In [3]:
# !pip install cvxopt

In [4]:
# Basic libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from pandas import read_csv, set_option
from pandas.plotting import scatter_matrix
import seaborn as sns
from datetime import date
import pkg_resources
import pip

# Import Model Packages 
import scipy.cluster.hierarchy as sch
from sklearn.cluster import AgglomerativeClustering
from scipy.cluster.hierarchy import fcluster
from scipy.cluster.hierarchy import dendrogram, linkage, cophenet
from sklearn.metrics import adjusted_mutual_info_score
from sklearn import cluster, covariance, manifold

# Package for optimization of mean variance optimization
import cvxopt as opt
from cvxopt import blas, solvers

# Machine Learning Package
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import RobustScaler
from xgboost import XGBRegressor
import shap

import warnings
warnings.filterwarnings('ignore')

Matplotlib created a temporary config/cache directory at /tmp/matplotlib-dt5ixg0y because the default path (/home/fgarcia/.cache/matplotlib) is not a writable directory; it is highly recommended to set the MPLCONFIGDIR environment variable to a writable directory, in particular to speed up the import of Matplotlib and to better support multiprocessing.
/opt/conda/lib/python3.9/site-packages/_distutils_hack/__init__.py:33: UserWarning: Setuptools is replacing distutils.
  warnings.warn("Setuptools is replacing distutils.")


In [5]:
# Figure caption
fig_num = 1


def fig_caption(title, caption):
    """Print figure caption on jupyter notebook"""
    global fig_num
    display(HTML(f"""<p style="font-size:11px;font-style:default;">
                     <center><b>
                     Figure {fig_num}. {title}.</b>
                     <br>{caption}</center></p>"""))
    fig_num += 1

***
<h2 style="color:#007CFE">Data Collection and Preprocessing</h2>

***

First, we load the dataset into a `pandas dataframe`.

In [6]:
# load dataset
dataset = pd.read_csv('SP500Data.csv', index_col=0)

# Display dataframe and shape
print("Table 02. Raw Data Extracted from Yahoo Finance.")
display(dataset)
print(f"   The number of rows is: {dataset.shape[0]}")
print(f"The number of columns is: {dataset.shape[1]}")

Table 02. Raw Data Extracted from Yahoo Finance.


ABT        ABBV        ABMD         ACN       ATVI  \
Date                                                                   
2018-01-02  58.790001   98.410004  192.490005  153.839996  64.309998   
2018-01-03  58.919998   99.949997  195.820007  154.550003  65.309998   
2018-01-04  58.820000   99.379997  199.250000  156.380005  64.660004   
2018-01-05  58.990002  101.110001  202.320007  157.669998  66.370003   
2018-01-08  58.820000   99.489998  207.800003  158.929993  66.629997   
...               ...         ...         ...         ...        ...   
2019-10-07  81.040001   74.330002  164.429993  186.809998  55.410000   
2019-10-08  78.510002   73.529999  159.490005  182.199997  54.130001   
2019-10-09  79.500000   73.300003  159.309998  184.339996  53.430000   
2019-10-10  80.139999   74.449997  162.339996  183.830002  53.689999   
2019-10-11  79.629997   73.629997  169.639999  184.970001  54.820000   

                  ADBE        AMD         AAP    AES         AMG  ...  \
Date                                                              ...   
2018-01-02  177.699997  10.980000  106.089996  10.88  203.039993  ...   
2018-01-03  181.039993  11.550000  107.050003  10.87  202.119995  ...   
2018-01-04  183.220001  12.120000  111.000000  10.83  198.539993  ...   
2018-01-05  185.339996  11.880000  112.180000  10.87  199.470001  ...   
2018-01-08  185.039993  12.280000  111.389999  10.87  200.529999  ...   
...                ...        ...         ...    ...         ...  ...   
2019-10-07  276.899994  28.930000  158.270004  15.85   74.830002  ...   
2019-10-08  270.829987  28.230000  154.330002  15.51   71.800003  ...   
2019-10-09  274.269989  28.459999  154.410004  15.54   72.250000  ...   
2019-10-10  274.980011  28.379999  155.949997  15.82   73.059998  ...   
2019-10-11  278.290009  29.750000  158.559998  16.01   75.970001  ...   

                  WLTW        WYNN        XEL        XRX       XLNX  \
Date                                                                  
2018-01-02  146.990005  164.300003  47.810001  29.370001  67.879997   
2018-01-03  149.740005  162.520004  47.490002  29.330000  69.239998   
2018-01-04  151.259995  163.399994  47.119999  29.690001  70.489998   
2018-01-05  152.229996  164.490005  46.790001  29.910000  74.150002   
2018-01-08  151.410004  162.300003  47.139999  30.260000  74.639999   
...                ...         ...        ...        ...        ...   
2019-10-07  187.589996  107.019997  64.279999  29.830000  92.629997   
2019-10-08  184.360001  102.570000  63.590000  28.309999  89.269997   
2019-10-09  186.199997  104.419998  63.919998  28.760000  90.480003   
2019-10-10  187.690002  106.059998  63.950001  28.820000  92.809998   
2019-10-11  188.100006  112.779999  63.330002  29.410000  96.220001   

                  XYL         YUM         ZBH       ZION         ZTS  
Date                                                                  
2018-01-02  68.070000   81.599998  124.059998  50.700001   71.769997  
2018-01-03  68.900002   81.529999  124.919998  50.639999   72.099998  
2018-01-04  69.360001   82.360001  124.739998  50.849998   72.529999  
2018-01-05  69.230003   82.839996  125.980003  50.869999   73.360001  
2018-01-08  69.480003   82.980003  126.220001  50.619999   74.239998  
...               ...         ...         ...        ...         ...  
2019-10-07  75.570000  113.690002  134.500000  43.439999  126.070000  
2019-10-08  73.430000  112.589996  131.220001  42.209999  125.379997  
2019-10-09  74.570000  113.330002  133.419998  42.419998  126.430000  
2019-10-10  75.430000  114.330002  134.070007  43.209999  127.410004  
2019-10-11  77.629997  114.449997  136.440002  44.090000  128.470001  

[448 rows x 502 columns]

   The number of rows is: 448
The number of columns is: 502


Show columns (equities) that have missing values.

In [7]:
missing_col = dataset.isna().sum()
display(missing_col[missing_col != 0])

CTVA    350
DOW     304
FOXA    298
FOX     299
dtype: int64

We will **drop** columns that have missing values of more than `30%` of their total.

In [8]:
missing_fractions = dataset.isnull().mean().sort_values(ascending=False)
missing_fractions.head(10)
drop_list = sorted(list(missing_fractions[missing_fractions > 0.3].index))
dataset.drop(labels=drop_list, axis=1, inplace=True)

Impute the remaining columns with null values using `forward fill`.

In [9]:
# Fill the missing values with the last value available in the dataset.
dataset = dataset.fillna(method='ffill')

# Display dataframe and shape
print("Table 03. Cleaned Data of S&P500 Stock Prices.")
display(dataset)
print(f"   The number of rows is: {dataset.shape[0]}")
print(f"The number of columns is: {dataset.shape[1]}")

Table 03. Cleaned Data of S&P500 Stock Prices.


ABT        ABBV        ABMD         ACN       ATVI  \
Date                                                                   
2018-01-02  58.790001   98.410004  192.490005  153.839996  64.309998   
2018-01-03  58.919998   99.949997  195.820007  154.550003  65.309998   
2018-01-04  58.820000   99.379997  199.250000  156.380005  64.660004   
2018-01-05  58.990002  101.110001  202.320007  157.669998  66.370003   
2018-01-08  58.820000   99.489998  207.800003  158.929993  66.629997   
...               ...         ...         ...         ...        ...   
2019-10-07  81.040001   74.330002  164.429993  186.809998  55.410000   
2019-10-08  78.510002   73.529999  159.490005  182.199997  54.130001   
2019-10-09  79.500000   73.300003  159.309998  184.339996  53.430000   
2019-10-10  80.139999   74.449997  162.339996  183.830002  53.689999   
2019-10-11  79.629997   73.629997  169.639999  184.970001  54.820000   

                  ADBE        AMD         AAP    AES         AMG  ...  \
Date                                                              ...   
2018-01-02  177.699997  10.980000  106.089996  10.88  203.039993  ...   
2018-01-03  181.039993  11.550000  107.050003  10.87  202.119995  ...   
2018-01-04  183.220001  12.120000  111.000000  10.83  198.539993  ...   
2018-01-05  185.339996  11.880000  112.180000  10.87  199.470001  ...   
2018-01-08  185.039993  12.280000  111.389999  10.87  200.529999  ...   
...                ...        ...         ...    ...         ...  ...   
2019-10-07  276.899994  28.930000  158.270004  15.85   74.830002  ...   
2019-10-08  270.829987  28.230000  154.330002  15.51   71.800003  ...   
2019-10-09  274.269989  28.459999  154.410004  15.54   72.250000  ...   
2019-10-10  274.980011  28.379999  155.949997  15.82   73.059998  ...   
2019-10-11  278.290009  29.750000  158.559998  16.01   75.970001  ...   

                  WLTW        WYNN        XEL        XRX       XLNX  \
Date                                                                  
2018-01-02  146.990005  164.300003  47.810001  29.370001  67.879997   
2018-01-03  149.740005  162.520004  47.490002  29.330000  69.239998   
2018-01-04  151.259995  163.399994  47.119999  29.690001  70.489998   
2018-01-05  152.229996  164.490005  46.790001  29.910000  74.150002   
2018-01-08  151.410004  162.300003  47.139999  30.260000  74.639999   
...                ...         ...        ...        ...        ...   
2019-10-07  187.589996  107.019997  64.279999  29.830000  92.629997   
2019-10-08  184.360001  102.570000  63.590000  28.309999  89.269997   
2019-10-09  186.199997  104.419998  63.919998  28.760000  90.480003   
2019-10-10  187.690002  106.059998  63.950001  28.820000  92.809998   
2019-10-11  188.100006  112.779999  63.330002  29.410000  96.220001   

                  XYL         YUM         ZBH       ZION         ZTS  
Date                                                                  
2018-01-02  68.070000   81.599998  124.059998  50.700001   71.769997  
2018-01-03  68.900002   81.529999  124.919998  50.639999   72.099998  
2018-01-04  69.360001   82.360001  124.739998  50.849998   72.529999  
2018-01-05  69.230003   82.839996  125.980003  50.869999   73.360001  
2018-01-08  69.480003   82.980003  126.220001  50.619999   74.239998  
...               ...         ...         ...        ...         ...  
2019-10-07  75.570000  113.690002  134.500000  43.439999  126.070000  
2019-10-08  73.430000  112.589996  131.220001  42.209999  125.379997  
2019-10-09  74.570000  113.330002  133.419998  42.419998  126.430000  
2019-10-10  75.430000  114.330002  134.070007  43.209999  127.410004  
2019-10-11  77.629997  114.449997  136.440002  44.090000  128.470001  

[448 rows x 498 columns]

   The number of rows is: 448
The number of columns is: 498


***
<h2 style="color:#007CFE">Exploratory Data Analysis</h2>

***

We will check the top and bottom stocks from this cleaned dataset.

In [10]:
df_returns = ((dataset - dataset.shift(+1)) / dataset.shift(+1)) * 100
df_returns = df_returns[1:]
print("Table 04. Stock Returns (%) of the Dataset.")
display(df_returns)
print(f"   The number of rows is: {df_returns.shape[0]}")
print(f"The number of columns is: {df_returns.shape[1]}")

Table 04. Stock Returns (%) of the Dataset.


ABT      ABBV      ABMD       ACN      ATVI      ADBE  \
Date                                                                     
2018-01-03  0.221121  1.564875  1.729961  0.461523  1.554968  1.879570   
2018-01-04 -0.169719 -0.570285  1.751605  1.184084 -0.995244  1.204158   
2018-01-05  0.289021  1.740796  1.540782  0.824909  2.644601  1.157076   
2018-01-08 -0.288188 -1.602218  2.708578  0.799134  0.391735 -0.161866   
2018-01-09  0.170008  0.753845  0.943211  0.333489 -0.660355  0.897105   
...              ...       ...       ...       ...       ...       ...   
2019-10-07 -1.158674 -0.442006 -2.847863 -1.492302 -0.054110 -0.021663   
2019-10-08 -3.121914 -1.076286 -3.004310 -2.467748 -2.310050 -2.192130   
2019-10-09  1.260983 -0.312792 -0.112865  1.174533 -1.293184  1.270170   
2019-10-10  0.805031  1.568887  1.901951 -0.276660  0.486615  0.258877   
2019-10-11 -0.636389 -1.101410  4.496737  0.620138  2.104677  1.203723   

                 AMD       AAP       AES       AMG  ...      WLTW      WYNN  \
Date                                                ...                       
2018-01-03  5.191263  0.904898 -0.091914 -0.453112  ...  1.870875 -1.083383   
2018-01-04  4.935062  3.689862 -0.367985 -1.771226  ...  1.015085  0.541465   
2018-01-05 -1.980196  1.063063  0.369344  0.468423  ...  0.641281  0.667082   
2018-01-08  3.367000 -0.704226  0.000000  0.531407  ... -0.538653 -1.331389   
2018-01-09 -3.745929 -0.807973 -1.011956 -0.129654  ...  0.277391  0.677752   
...              ...       ...       ...       ...  ...       ...       ...   
2019-10-07 -0.275767 -1.044140 -1.122896 -2.144631  ... -0.561886 -1.735384   
2019-10-08 -2.419636 -2.489418 -2.145111 -4.049176  ... -1.721838 -4.158099   
2019-10-09  0.814734  0.051838  0.193422  0.626737  ...  0.998045  1.803645   
2019-10-10 -0.281096  0.997340  1.801800  1.121104  ...  0.800218  1.570580   
2019-10-11  4.827346  1.673614  1.201015  3.983033  ...  0.218447  6.336038   

                 XEL       XRX      XLNX       XYL       YUM       ZBH  \
Date                                                                     
2018-01-03 -0.669315 -0.136197  2.003537  1.219336 -0.085784  0.693213   
2018-01-04 -0.779117  1.227414  1.805315  0.667633  1.018032 -0.144092   
2018-01-05 -0.700335  0.740988  5.192231 -0.187424  0.582802  0.994072   
2018-01-08  0.748020  1.170178  0.660820  0.361115  0.169009  0.190505   
2018-01-09 -1.166736 -0.231327  0.267949  0.028780 -0.265126 -1.608302   
...              ...       ...       ...       ...       ...       ...   
2019-10-07 -0.541542  0.811084 -1.708405 -0.526526 -0.845980  0.037191   
2019-10-08 -1.073427 -5.095543 -3.627335 -2.831811 -0.967549 -2.438661   
2019-10-09  0.518946  1.589547  1.355446  1.552498  0.657257  1.676571   
2019-10-10  0.046938  0.208621  2.575148  1.153280  0.882379  0.487190   
2019-10-11 -0.969506  2.047190  3.674177  2.916607  0.104955  1.767730   

                ZION       ZTS  
Date                            
2018-01-03 -0.118346  0.459805  
2018-01-04  0.414690  0.596394  
2018-01-05  0.039332  1.144357  
2018-01-08 -0.491449  1.199560  
2018-01-09  2.350854  1.171879  
...              ...       ...  
2019-10-07  0.555551 -0.771352  
2019-10-08 -2.831491 -0.547317  
2019-10-09  0.497510  0.837457  
2019-10-10  1.862331  0.775135  
2019-10-11  2.036568  0.831958  

[447 rows x 498 columns]

   The number of rows is: 447
The number of columns is: 498


In [11]:
df_top = df_returns.mean(axis=0).to_frame().sort_values(by=0, ascending=False)
df_top.columns = ['Returns %']
df_top10 = df_top.head(10)

In [12]:
plt.figure(figsize=(10, 6))
sns.barplot(x=df_top10['Returns %'], y=df_top10.index,
            orient='h', color='#0531BC')
plt.title('Top 10 Stocks Average Daily Return %', fontsize=15)
plt.xlabel('Returns %')
plt.ylabel('Stocks')

# Save the plot as a PNG file
plt.savefig("Top10_Daily_Returns.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="Top10_Daily_Returns.png" alt="plots" style="display:'
           f'block;margin-left:auto;margin-right:auto;width:60%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()
fig_caption('Top 10 Stocks Average Daily Return %', '')

These **top 10 stocks**, in terms of *average daily return %*, are the **ideal stocks** to look into for the long term. This top includes the following stocks: Advanced Micro Devices, Inc. (`AMD`), Chipotle Mexican Grill, Inc. (`CMG`), and Keysight Technologies Inc. (`KEYS`). However, one must take into consideration the volatility of these returns.

In [13]:
df_bot10 = df_top.tail(10)
plt.figure(figsize=(10, 6))
sns.barplot(x=df_bot10['Returns %'], y=df_bot10.index,
            orient='h', color='#0531BC')
plt.title('Bottom 10 Stocks Average Daily Return %', fontsize=15)
plt.xlabel('Returns %')
plt.ylabel('Stocks')

# Save the plot as a PNG file
plt.savefig("Bottom10_Daily_Returns.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="Bottom10_Daily_Returns.png" alt="plots" style="display'
           f':block;margin-left:auto;margin-right:auto;width:60%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()
fig_caption('Bottom 10 Stocks Average Daily Return %', '')

These **bottom 10 stocks**, in terms of *average daily return %*, are the **NOT ideal stocks** to look into for the long term. This top includes the following stocks: Core Lithium Ltd. (`CXO`), Schlumberger NV (`SLB`), and Invesco Ltd. (`IVZ`). Again, volatility plays a key role in asset selection.

<h1 style="background-color:#0531BC; color:#F5F5F1; padding: 20px;">Results and Discussion</h1>

***
<h2 style="color:#007CFE">Train - Test Split</h2>

***

For the purpose of clustering, we will be using annual returns. Additionally, we will train the data, followed by testing. Let us prepare the dataset for training and testing by separating 20% of the dataset for testing, followed by generating the return series.

In [14]:
# Split dataset
X = dataset.copy('deep')
row = len(X)
train_len = int(row*.8)
X_train = X.head(train_len)
X_test = X.tail(row-train_len)

# Calculate percentage return
returns = X_train.pct_change().dropna()
returns_test = X_test.pct_change().dropna()

# Display dataframe and shape
print("Table 05. X train set.")
display(returns)
print(f"   X train set number of rows is: {returns.shape[0]}")
print(f"X train set number of columns is: {returns.shape[1]}")

Table 05. X train set.


ABT      ABBV      ABMD       ACN      ATVI      ADBE  \
Date                                                                     
2018-01-03  0.002211  0.015649  0.017300  0.004615  0.015550  0.018796   
2018-01-04 -0.001697 -0.005703  0.017516  0.011841 -0.009952  0.012042   
2018-01-05  0.002890  0.017408  0.015408  0.008249  0.026446  0.011571   
2018-01-08 -0.002882 -0.016022  0.027086  0.007991  0.003917 -0.001619   
2018-01-09  0.001700  0.007538  0.009432  0.003335 -0.006604  0.008971   
...              ...       ...       ...       ...       ...       ...   
2019-05-30  0.011233 -0.005765  0.023121  0.006813 -0.009775  0.004133   
2019-05-31 -0.005097 -0.011596 -0.016780 -0.004194 -0.004362 -0.013151   
2019-06-03 -0.005517 -0.013166 -0.001871 -0.016342 -0.022827 -0.043817   
2019-06-04  0.023115  0.013871  0.021535  0.016042  0.029023  0.037370   
2019-06-05  0.015879  0.004039  0.006515  0.008934  0.020867  0.015444   

                 AMD       AAP       AES       AMG  ...      WLTW      WYNN  \
Date                                                ...                       
2018-01-03  0.051913  0.009049 -0.000919 -0.004531  ...  0.018709 -0.010834   
2018-01-04  0.049351  0.036899 -0.003680 -0.017712  ...  0.010151  0.005415   
2018-01-05 -0.019802  0.010631  0.003693  0.004684  ...  0.006413  0.006671   
2018-01-08  0.033670 -0.007042  0.000000  0.005314  ... -0.005387 -0.013314   
2018-01-09 -0.037459 -0.008080 -0.010120 -0.001297  ...  0.002774  0.006778   
...              ...       ...       ...       ...  ...       ...       ...   
2019-05-30 -0.002136  0.023174 -0.010585 -0.013300  ...  0.011384 -0.025876   
2019-05-31 -0.022119 -0.027543 -0.005664 -0.026028  ... -0.012325 -0.036794   
2019-06-03  0.006202 -0.011806  0.008228  0.007039  ...  0.014587 -0.035032   
2019-06-04  0.072154  0.009401  0.009416  0.035422  ...  0.006515  0.089794   
2019-06-05 -0.002367  0.000000  0.025498  0.007551  ...  0.020254 -0.021352   

                 XEL       XRX      XLNX       XYL       YUM       ZBH  \
Date                                                                     
2018-01-03 -0.006693 -0.001362  0.020035  0.012193 -0.000858  0.006932   
2018-01-04 -0.007791  0.012274  0.018053  0.006676  0.010180 -0.001441   
2018-01-05 -0.007003  0.007410  0.051922 -0.001874  0.005828  0.009941   
2018-01-08  0.007480  0.011702  0.006608  0.003611  0.001690  0.001905   
2018-01-09 -0.011667 -0.002313  0.002679  0.000288 -0.002651 -0.016083   
...              ...       ...       ...       ...       ...       ...   
2019-05-30  0.002797  0.000317  0.017366  0.002163  0.019067  0.005970   
2019-05-31 -0.000349 -0.030101 -0.029777  0.001214  0.007878  0.009213   
2019-06-03  0.010638  0.012088 -0.000098  0.020210  0.017196  0.003248   
2019-06-04 -0.002934  0.074887  0.044966  0.022187  0.027471  0.027209   
2019-06-05  0.026653  0.003604 -0.012161  0.019121  0.002991  0.009624   

                ZION       ZTS  
Date                            
2018-01-03 -0.001183  0.004598  
2018-01-04  0.004147  0.005964  
2018-01-05  0.000393  0.011444  
2018-01-08 -0.004914  0.011996  
2018-01-09  0.023509  0.011719  
...              ...       ...  
2019-05-30 -0.014571  0.011666  
2019-05-31 -0.020246 -0.012509  
2019-06-03  0.005805  0.028699  
2019-06-04  0.024931  0.040115  
2019-06-05 -0.004955  0.003515  

[357 rows x 498 columns]

   X train set number of rows is: 357
X train set number of columns is: 498


In [15]:
# Display dataframe and shape
print("Table 06. X test set.")
display(returns_test)
print(f"   X test set number of rows is: {returns_test.shape[0]}")
print(f"X test set number of columns is: {returns_test.shape[1]}")

Table 06. X test set.


ABT      ABBV      ABMD       ACN      ATVI      ADBE  \
Date                                                                     
2019-06-07  0.008116  0.004671 -0.004904  0.013969  0.029492  0.012227   
2019-06-10  0.006564 -0.006199  0.017023  0.008310  0.013883  0.007837   
2019-06-11 -0.001846  0.015854 -0.038175 -0.005801  0.014997 -0.015517   
2019-06-12  0.010232  0.005117 -0.028585  0.006653 -0.034475  0.003080   
2019-06-13  0.004149  0.004836  0.003889  0.001950  0.021291 -0.001951   
...              ...       ...       ...       ...       ...       ...   
2019-10-07 -0.011587 -0.004420 -0.028479 -0.014923 -0.000541 -0.000217   
2019-10-08 -0.031219 -0.010763 -0.030043 -0.024677 -0.023101 -0.021921   
2019-10-09  0.012610 -0.003128 -0.001129  0.011745 -0.012932  0.012702   
2019-10-10  0.008050  0.015689  0.019020 -0.002767  0.004866  0.002589   
2019-10-11 -0.006364 -0.011014  0.044967  0.006201  0.021047  0.012037   

                 AMD       AAP       AES       AMG  ...      WLTW      WYNN  \
Date                                                ...                       
2019-06-07  0.018542  0.002905 -0.015330  0.003033  ...  0.018586  0.009456   
2019-06-10  0.025301 -0.011780 -0.016168  0.010081  ...  0.007652  0.050892   
2019-06-11 -0.024676 -0.008794 -0.003043 -0.001885  ...  0.000000  0.012685   
2019-06-12 -0.007097 -0.010712  0.023199 -0.008444  ... -0.001062 -0.028862   
2019-06-13 -0.024549  0.000465  0.005967  0.005714  ...  0.003190  0.021527   
...              ...       ...       ...       ...  ...       ...       ...   
2019-10-07 -0.002758 -0.010441 -0.011229 -0.021446  ... -0.005619 -0.017354   
2019-10-08 -0.024196 -0.024894 -0.021451 -0.040492  ... -0.017218 -0.041581   
2019-10-09  0.008147  0.000518  0.001934  0.006267  ...  0.009980  0.018036   
2019-10-10 -0.002811  0.009973  0.018018  0.011211  ...  0.008002  0.015706   
2019-10-11  0.048273  0.016736  0.012010  0.039830  ...  0.002184  0.063360   

                 XEL       XRX      XLNX       XYL       YUM       ZBH  \
Date                                                                     
2019-06-07 -0.006187  0.003820  0.013961  0.005180  0.005995  0.003491   
2019-06-10 -0.002861  0.015515  0.031538  0.010307 -0.003851  0.008117   
2019-06-11 -0.009619  0.009513  0.007125 -0.004728 -0.009940 -0.010106   
2019-06-12  0.014142 -0.003141 -0.033402  0.002625  0.012736  0.001743   
2019-06-13 -0.001176  0.007448  0.003798  0.010971  0.000367 -0.007457   
...              ...       ...       ...       ...       ...       ...   
2019-10-07 -0.005415  0.008111 -0.017084 -0.005265 -0.008460  0.000372   
2019-10-08 -0.010734 -0.050955 -0.036273 -0.028318 -0.009675 -0.024387   
2019-10-09  0.005189  0.015895  0.013554  0.015525  0.006573  0.016766   
2019-10-10  0.000469  0.002086  0.025751  0.011533  0.008824  0.004872   
2019-10-11 -0.009695  0.020472  0.036742  0.029166  0.001050  0.017677   

                ZION       ZTS  
Date                            
2019-06-07 -0.013562  0.010745  
2019-06-10  0.004583  0.001454  
2019-06-11  0.013230 -0.000544  
2019-06-12 -0.005628  0.011619  
2019-06-13  0.012678 -0.012653  
...              ...       ...  
2019-10-07  0.005556 -0.007714  
2019-10-08 -0.028315 -0.005473  
2019-10-09  0.004975  0.008375  
2019-10-10  0.018623  0.007751  
2019-10-11  0.020366  0.008320  

[89 rows x 498 columns]

   X test set number of rows is: 89
X test set number of columns is: 498


***
<h2 style="color:#007CFE">Mean-Variance Portfolio (MVP) Construction</h2>

***

To construct the `MVP` we will be using the `cvxopt` library to do the optimization.

In [16]:
def constructMVP(cov):
    """Construct MVP using optimization."""
    cov = cov.T.values
    n = len(cov)
    N = 100
    mus = [10 ** (5.0 * t / N - 1.0) for t in range(N)]

    # Convert to cvxopt matrices
    S = opt.matrix(cov)

    pbar = opt.matrix(np.ones(cov.shape[0]))

    # Create constraint matrices
    G = -opt.matrix(np.eye(n))  # negative n x n identity matrix
    h = opt.matrix(0.0, (n, 1))
    A = opt.matrix(1.0, (1, n))
    b = opt.matrix(1.0)

    # Calculate efficient frontier weights using quadratic programming
    solvers.options['show_progress'] = False
    portfolios = [solvers.qp(mu * S, -pbar, G, h, A, b)['x']
                  for mu in mus]

    # CALCULATE RISKS AND RETURNS FOR FRONTIER
    returns = [blas.dot(pbar, x) for x in portfolios]
    risks = [np.sqrt(blas.dot(x, S * x)) for x in portfolios]

    # CALCULATE THE 2ND DEGREE POLYNOMIAL OF THE FRONTIER CURVE
    m1 = np.polyfit(returns, risks, 2)
    x1 = np.sqrt(m1[2] / m1[0])

    # CALCULATE THE OPTIMAL PORTFOLIO
    wt = solvers.qp(opt.matrix(x1 * S), -pbar, G, h, A, b)['x']

    return list(wt)

In [17]:
# Returns covariance
cov = returns.cov()

# MVP Portfolio
MVP_port = pd.DataFrame(constructMVP(cov), index=cov.index)
top_10_MVP = MVP_port.sort_values(by=0, ascending=False)[:10]

In [18]:
# Display the results
fig, ax = plt.subplots(figsize=(10, 6))
top_10_MVP.sort_values(by=0, ascending=True).plot(kind='barh', ax=ax,
                                                  color='#0531BC')
ax.set_title('Top 10 Stocks in MVP', size=15)
ax.set_ylabel('Stocks Symbol')
ax.set_xlabel('Weight in Portfolio')
ax.get_legend().remove()

# Save the plot as a PNG file
plt.savefig("Top10_MVP.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="Top10_MVP.png" alt="plots" style="display:block;'
           f'margin-left:auto;margin-right:auto;width:60%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()
fig_caption('Top 10 Stocks in MVP', '')

`MVP` allocated its investments across different stocks. As a result, the stock with the highest weight includes the following stocks: Amcor PLC (`AMCR`) - `14.81%` , Duke Energy Corp (`DUK`) - `10.23%`, and Newmont Corporation (`NEM`) - `7.87%`.

***
<h2 style="color:#007CFE">Hierarchical Risk Parity (HRP) Portfolio Construction</h2>

***

For the `HRP` portfolio construction we will follow the following steps:
* Hierarchical Clustering
* Matrix Seriation
* Recursive Bisection
* Getting of Portfolio Weights for all Types of Allocation

<h3 style="color:#221F1F">Hierarchical Clustering</h3>

***

In [19]:
def correlDist(corr):
    """Transform correlation matrix into correlation distance using 
    Gower metric."""
    dist = ((1 - corr) / 2.)**.5
    return dist

In [20]:
# Calulate linkage
dist = correlDist(returns.corr())
link = linkage(dist, 'ward')

In [21]:
# Plot Dendrogram
plt.figure(figsize=(12, 8))
dendrogram(link, labels=X.columns)
plt.xticks([], [])
plt.title("Dendrogram of Equity Assets")
plt.xlabel('Assets')
plt.ylabel('Distance between clusters')

# Save the plot as a PNG file
plt.savefig("Dendrogram.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="Dendrogram.png" alt="plots" style="display:block;'
           f'margin-left:auto;margin-right:auto;width:80%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()
fig_caption('Dendrogram of Equity Assets', '')

After doing `Hierarchical clustering` the dendrogram above reflects the clustering of different assets.

<h3 style="color:#221F1F">Matrix Seriation</h3>

***

In [22]:
def matrix_diag(link):
    """Sort clustered items by distance."""
    link = link.astype(int)
    sortIx = pd.Series([link[-1, 0], link[-1, 1]])
    numItems = link[-1, 3]  # number of original items
    while sortIx.max() >= numItems:
        sortIx.index = range(0, sortIx.shape[0] * 2, 2)  # make space
        df0 = sortIx[sortIx >= numItems]  # find clusters
        i = df0.index
        j = df0.values - numItems
        sortIx[i] = link[j, 0]  # item 1
        df0 = pd.Series(link[j, 1], index=i + 1)
        sortIx = sortIx.append(df0)  # item 2
        sortIx = sortIx.sort_index()  # re-sort
        sortIx.index = range(sortIx.shape[0])  # re-index
    return sortIx.tolist()

In [23]:
# Get stocks
corr = returns.corr()
print("Get the stocks (shown only are the first 20 stocks in the list):")
print(returns.corr().index[matrix_diag(link)].tolist()[:20])

Get the stocks (shown only are the first 20 stocks in the list):
['CMS', 'WEC', 'XEL', 'LNT', 'AEE', 'DUK', 'ED', 'AEP', 'DTE', 'EVRG', 'NI', 'D', 'SO', 'FE', 'PPL', 'CNP', 'AWK', 'ATO', 'EXC', 'PEG']


<h3 style="color:#221F1F">Recursive Bisection</h3>

***

In [24]:
def constructIVP(cov, **kargs):
    """Compute the inverse-variance portfolio."""
    ivp = 1. / np.diag(cov)
    ivp /= ivp.sum()
    return ivp


def getClusterVar(cov, cItems):
    """Compute variance per cluster."""
    cov_ = cov.loc[cItems, cItems]  # matrix slice
    w_ = constructIVP(cov_).reshape(-1, 1)
    cVar = np.dot(np.dot(w_.T, cov_), w_)[0, 0]
    return cVar


def recursive_bisection(cov, sortIx):
    """Compute HRP allocation."""
    w = pd.Series(1, index=sortIx)
    cItems = [sortIx]  # initialize all items in one cluster
    while len(cItems) > 0:
        cItems = (
            [i[j:k] for i in cItems for j, k in (
                (0, len(i) // 2),
                (len(i) // 2, len(i))
            ) if len(i) > 1]  # bi-section
        )

        for i in range(0, len(cItems), 2):  # parse in pairs
            cItems0 = cItems[i]  # cluster 1
            cItems1 = cItems[i + 1]  # cluster 2
            cVar0 = getClusterVar(cov, cItems0)
            cVar1 = getClusterVar(cov, cItems1)
            alpha = 1 - cVar0 / (cVar0 + cVar1)
            w[cItems0] *= alpha  # update weight 1
            w[cItems1] *= 1 - alpha  # update weight 2

    return w

In [25]:
# HRP weights
sortIx = returns.corr().index[matrix_diag(link)].tolist()

In [26]:
def constructIVP(cov, **kargs):
    """Compute the inverse-variance portfolio."""
    ivp = 1. / np.diag(cov)
    ivp /= ivp.sum()
    return ivp


def getHRP(cov, corr):
    """Construct a hierarchical portfolio."""
    dist = correlDist(corr)
    link = sch.linkage(dist, 'single')
    sortIx = matrix_diag(link)
    sortIx = corr.index[sortIx].tolist()
    hrp = recursive_bisection(cov, sortIx)
    return hrp.sort_index()

In [27]:
# Get the top 10 of HRP portfolio
HRP_port = pd.DataFrame(getHRP(cov, corr), index=cov.index)
top_10_HRP = HRP_port.sort_values(by=0, ascending=False)[:10]

In [28]:
# Display the results
fig, ax = plt.subplots(figsize=(10, 6))
top_10_HRP.sort_values(by=0, ascending=True).plot(kind='barh', ax=ax,
                                                  color='#0531BC')
ax.set_title('Top 10 Stocks in HRP Portfolio', size=15)
ax.set_ylabel('Stocks Symbol')
ax.set_xlabel('Weight in Portfolio')
ax.get_legend().remove()

# Save the plot as a PNG file
plt.savefig("Top10_HRP.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="Top10_HRP.png" alt="plots" style="display:block;'
           f'margin-left:auto;margin-right:auto;width:60%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()
fig_caption('Top 10 Stocks in HRP Portfolio', '')

`HRP` portfolio allocated its investments across different stocks. As a result, the stock with the highest weight includes the following stocks: Public Service Enterprise Group Inc. (`PEG`) - `1.31%` , Amcor PLC (`AMCR`) - `1.21%`, and CME Group Inc (`CME`) - `1.13%`.

***
<h2 style="color:#007CFE">Portfolio Summary</h2>

***

In [29]:
def get_all_portfolios(returns):
    """Construct the MVP and HRP portfolio."""
    cov, corr = returns.cov(), returns.corr()
    hrp = getHRP(cov, corr)
    mvp = constructMVP(cov)
    mvp = pd.Series(mvp, index=cov.index)
    portfolios = pd.DataFrame([mvp, hrp], index=['MVP', 'HRP']).T

    return portfolios

In [30]:
# Calculate the portfolio weights of MVP and HRP
portfolios = get_all_portfolios(returns)

# Display dataframe and shape
print("Table 07. Weights of each Stock for MVP and HRP.")
display(portfolios)
print(f"   Portfolio number of rows is: {portfolios.shape[0]}")
print(f"Portfolio number of columns is: {portfolios.shape[1]}")

Table 07. Weights of each Stock for MVP and HRP.


MVP       HRP
ABT   0.000039  0.001364
ABBV  0.000039  0.000967
ABMD  0.000029  0.000625
ACN   0.000050  0.001253
ATVI  0.000084  0.000696
...        ...       ...
XYL   0.000047  0.001129
YUM   0.000378  0.004225
ZBH   0.000052  0.001659
ZION  0.000081  0.001053
ZTS   0.000043  0.001687

[498 rows x 2 columns]

   Portfolio number of rows is: 498
Portfolio number of columns is: 2


The row here represents the equity stock of a company while the column corresponds to the portfolio strategy (`MVP` or `HRP`). The values are the weights for the respective portfolio strategy.

In [31]:
# Get the number of assets with at least 0.1% weight
count_MVP = len(list(filter(lambda x: x > 0.001, constructMVP(cov))))
count_HRP = len(list(filter(lambda x: x > 0.001, getHRP(cov, corr))))

In [32]:
# Display the count of assets per portfolio strategy
fig, ax = plt.subplots(figsize=(10, 6))
ax.bar(['HRP', 'MVP'], [count_HRP, count_MVP], color='#0531BC')
ax.set_title('Count of Stock with at least 0.1% Weight', size=15)
ax.set_ylabel('Stock Count')
ax.set_xlabel('Portfolio Strategy')

# Save the plot as a PNG file
plt.savefig("Stock_Count.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="Stock_Count.png" alt="plots" style="display:block;'
           f'margin-left:auto;margin-right:auto;width:60%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()
fig_caption('Count of Stock with at least 0.1% Weight', '')

**Diversification** as a main strength is highlighted here in the portfolio composition overview. In terms of total assets included in the portfolio, `MVP` is concentrated on `46` stocks, while the `HRP` portfolio is better spread out at `388`.

In [33]:
MVP_HRP_compare = top_10_MVP.copy()
MVP_HRP_compare['HRP'] = portfolios.loc[top_10_MVP.index, 'HRP']
MVP_HRP_compare = MVP_HRP_compare.rename(columns={0: 'MVP'})
MVP_HRP_compare = pd.concat(
    [MVP_HRP_compare['HRP'], MVP_HRP_compare['MVP']], axis=1)
MVP_HRP_compare
fig, ax = plt.subplots(figsize=(10, 6))
MVP_HRP_compare.sort_values(by='MVP', ascending=True).plot(
    kind='barh', ax=ax, color=['#007CFE', '#0531BC'])
ax.set_title('Comparison of Portfolio Weights of MVP and HRP', size=15)
ax.set_ylabel('Stock Symbol')
ax.set_xlabel('Stock Weight')

# Save the plot as a PNG file
plt.savefig("MVP_HRP_Compare.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="MVP_HRP_Compare.png" alt="plots" style="display:block;'
           f'margin-left:auto;margin-right:auto;width:70%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()
fig_caption('Comparison of Portfolio Weights of MVP and HRP', '')

Based on the top ten stocks with the highest weights, we notice MVP allocates only to 5 industries (with the highest weight at a significant 25%) versus HRP, which allocates more evenly across industries based on risk profile.

***
<h2 style="color:#007CFE">Portfolio Comparison</h2>

***

<h3 style="color:#221F1F">In-Sample Comparison</h3>

***

In [34]:
Insample_Result = pd.DataFrame(np.dot(returns, np.array(portfolios)),
                               columns=['MVP', 'HRP'], index=returns.index)
OutOfSample_Result = pd.DataFrame(np.dot(returns_test, np.array(portfolios)),
                                  columns=['MVP', 'HRP'],
                                  index=returns_test.index)

In [35]:
Insample_Result.cumsum().plot(figsize=(13, 5), title="In-Sample Results",
                              color=['#0531BC', '#007CFE'])
plt.ylabel('Cummulative Returns')

# Save the plot as a PNG file
plt.savefig("InSample_Res.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="InSample_Res.png" alt="plots" style="display:block;'
           f'margin-left:auto;margin-right:auto;width:90%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()
fig_caption('In-Sample Results', '')

In [36]:
# In_sample Results
stddev = Insample_Result.std() * np.sqrt(252)
sharp_ratio = (Insample_Result.mean()*np.sqrt(252))/(Insample_Result).std()
Results = pd.DataFrame(dict(stdev=stddev, sharp_ratio=sharp_ratio))
print("Table 08. In-Sample standard deviation and Sharpe Ratio Results.")
display(Results)

Table 08. In-Sample standard deviation and Sharpe Ratio Results.


stdev  sharp_ratio
MVP  0.085516     0.785019
HRP  0.126944     0.523599

*Note: Standard Deviation comparisons shown were made using simple standard deviations of individual securities.*

Visually, the light-blue line representing the **HRP outperforms**, at least at a glance, the Mean-Variance Portfolio for most of the periods in the training data. Comparisons of cumulative returns see the HRP portfolio above the MVP. However, comparisons of Sharpe Ratios tell a different story: the *MVP portfolio provides a higher Sharpe Ratio than HRP*. Why does the Sharpe ratio provide results opposite to the intuition seen in the graph? Several factors may be at play: 1) The concept of **risk-adjusted returns** used in the Sharpe Ratio is most similar to the concepts used in the mean-variance portfolio, and 2) **asset volatility** here is computed in a naïve manner versus the standard measure of computing portfolio volatility.

<h3 style="color:#221F1F">Out-Of-Sample Results</h3>

***

In [37]:
OutOfSample_Result.cumsum().plot(figsize=(13, 5),
                                 title="Out-Of-Sample Results",
                                 color=['#0531BC', '#007CFE'])
plt.ylabel('Cummulative Returns')

# Save the plot as a PNG file
plt.savefig("OutSample_Res.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="OutSample_Res.png" alt="plots" style="display:block;'
           f'margin-left:auto;margin-right:auto;width:90%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()
fig_caption('Out-Of-Sample Results', '')

In [38]:
# Outof_sample Results
stddev_oos = OutOfSample_Result.std() * np.sqrt(252)
sharp_ratio_oos = ((OutOfSample_Result.mean()*np.sqrt(252))
                   / (OutOfSample_Result).std())
Results_oos = pd.DataFrame(
    dict(stdev_oos=stddev_oos, sharp_ratio_oos=sharp_ratio_oos))
print("Table 09. Out-of-Sample standard deviation and Sharpe Ratio Results.")
display(Results)

Table 09. Out-of-Sample standard deviation and Sharpe Ratio Results.


stdev  sharp_ratio
MVP  0.085516     0.785019
HRP  0.126944     0.523599

A key attribute proven empirically is that **HRP portfolios perform better than MVP on out-of-sample tests**; running the portfolio on the test period reinforces this claim. Sharpe Ratio is higher in this case for the HRP portfolio. As established earlier, in periods of *higher volatility, the primary strength of HRP is providing more stable returns due to diversification*.

So, what now? For most investors, a comparison of Sharpe Ratios is enough. But we are not most people. Access to technologies such as explainable AI allows us to compare these two strategies more granularly and identify key assets and their corresponding statistics.

***
<h2 style="color:#007CFE">Data Preparation for Explainable AI</h2>

***

In [39]:
def clean_df_for_ml(df):
    """Setting-up the dataset of derived statistical features from the stock
    returns data of S&P500."""
    df['Date'] = pd.to_datetime(df['Date'])
    df = df.set_index('Date')

    # add statitistics for each asset_i
    monthly_df = (
        df.groupby(pd.Grouper(freq='M')).agg(
            {col: [
                ('mean_' + col, 'mean'),
                ('std_' + col, 'std'),
                ('min-max_' + col, lambda x: x.max() - x.min())
            ] for col in df.columns})
    )
    monthly_df = monthly_df.reset_index()
    monthly_df.columns = monthly_df.columns.droplevel()
    monthly_df = monthly_df.rename(columns={'': 'Month'})

    # Add mean of asset class
    monthly_df['mean_mean'] = (df.groupby(pd.Grouper(freq='M'))
                               .apply(lambda x: x.mean().mean()).values)

    # Add std of asset returns
    monthly_df['mean_std'] = (df.groupby(pd.Grouper(freq='M'))
                              .apply(lambda x: x.mean().std()).values)

    # Add mean of drawdown
    monthly_df['mean_dd'] = (df.groupby(pd.Grouper(freq='M'))
                             .apply(lambda x: (x.max() - x.min())
                                    .mean()).values)
    # Correlation
    monthly_df['mean_corr'] = (df.groupby(pd.Grouper(freq='M'))
                               .apply(lambda x: x.corr().mean().mean())
                               .values)

    # Add running stats
    monthly_df['rolling_mean'] = df.groupby(pd.Grouper(
        freq='M')).apply(lambda x: x.mean().mean()).values
    rmeans = []
    for i in range(len(monthly_df)):
        rmean = np.std(monthly_df['rolling_mean'][:i+1])
        rmeans.append(rmean)
    monthly_df['rolling_mean'] = rmeans

    monthly_df['rolling_std'] = df.groupby(pd.Grouper(
        freq='M')).apply(lambda x: x.mean().std()).values
    rstds = []
    for i in range(len(monthly_df)):
        rstd = np.std(monthly_df['rolling_std'][:i+1])
        rstds.append(rstd)
    monthly_df['rolling_std'] = rstds

    monthly_df['rolling_corr'] = (df.groupby(pd.Grouper(freq='M'))
                                  .apply(lambda x: x.corr().mean().mean())
                                  .values)
    rcorrs = []
    for i in range(len(monthly_df)):
        rcorr = np.std(monthly_df['rolling_corr'][:i+1])
        rcorrs.append(rcorr)
    monthly_df['rolling_corr'] = rcorrs

    monthly_df['rolling_dd'] = (df.groupby(pd.Grouper(freq='M'))
                                .apply(lambda x: (x.max() - x.min())
                                       .mean()).values)
    rdds = []
    for i in range(len(monthly_df)):
        rdd = np.std(monthly_df['rolling_dd'][:i+1])
        rdds.append(rdd)
    monthly_df['rolling_dd'] = rdds

    return monthly_df

In [40]:
ret2 = returns.copy()
ret2 = ret2.reset_index()
ret3 = clean_df_for_ml(ret2)
ret3 = ret3.set_index('Month')

# Display dataframe and shape
print("Table 10. Data with statistics on stocks for XAI Analysis.")
display(ret3)
print(f"   Prepared dataset number of rows is: {ret3.shape[0]}")
print(f"Prepared dataset number of columns is: {ret3.shape[1]}")

Table 10. Data with statistics on stocks for XAI Analysis.


mean_ABT   std_ABT  min-max_ABT  mean_ABBV  std_ABBV  \
Month                                                              
2018-01-31  0.002856  0.011830     0.055781   0.007175  0.036130   
2018-02-28 -0.001420  0.017721     0.068915   0.001947  0.024170   
2018-03-31 -0.000209  0.015606     0.053488  -0.009062  0.031601   
2018-04-30 -0.001335  0.015011     0.054058   0.001168  0.021689   
2018-05-31  0.002657  0.012082     0.043392   0.001315  0.020717   
2018-06-30 -0.000392  0.007595     0.033191  -0.003066  0.010901   
2018-07-31  0.003496  0.011582     0.051349  -0.000027  0.019820   
2018-08-31  0.000878  0.006995     0.025804   0.001799  0.011595   
2018-09-30  0.004966  0.010887     0.051153  -0.000683  0.013786   
2018-10-31 -0.002580  0.015722     0.065996  -0.008215  0.020998   
2018-11-30  0.003551  0.017190     0.075393   0.009340  0.019819   
2018-12-31 -0.001013  0.021838     0.099224  -0.000965  0.021096   
2019-01-31  0.000658  0.021981     0.079419  -0.006228  0.026174   
2019-02-28  0.003286  0.008847     0.030874  -0.000644  0.010099   
2019-03-31  0.001453  0.010134     0.041504   0.000851  0.009885   
2019-04-30 -0.000129  0.014264     0.067163  -0.000647  0.011822   
2019-05-31 -0.001955  0.009839     0.045168  -0.001508  0.010364   
2019-06-30  0.011159  0.014888     0.028631   0.001581  0.013685   

            min-max_ABBV  mean_ABMD  std_ABMD  min-max_ABMD  mean_ACN  ...  \
Month                                                                  ...   
2018-01-31      0.190244   0.010081  0.010656      0.037072  0.002209  ...   
2018-02-28      0.085858   0.007514  0.034190      0.131628  0.000227  ...   
2018-03-31      0.152175   0.004058  0.018590      0.070543 -0.001993  ...   
2018-04-30      0.094332   0.001827  0.021668      0.083031 -0.000602  ...   
2018-05-31      0.096683   0.011167  0.028520      0.137568  0.001385  ...   
2018-06-30      0.039471   0.003575  0.020659      0.076036  0.002466  ...   
2018-07-31      0.072965  -0.006368  0.028986      0.134612 -0.001224  ...   
2018-08-31      0.049473   0.006106  0.016633      0.069118  0.002597  ...   
2018-09-30      0.062080   0.005911  0.035187      0.157893  0.000369  ...   
2018-10-31      0.076451  -0.011481  0.030794      0.109524 -0.003162  ...   
2018-11-30      0.070890   0.000234  0.053713      0.270876  0.002127  ...   
2018-12-31      0.095292  -0.000572  0.037232      0.148875 -0.007849  ...   
2019-01-31      0.094397   0.004040  0.027704      0.102861  0.004163  ...   
2019-02-28      0.040599  -0.002304  0.022300      0.100854  0.002639  ...   
2019-03-31      0.035603  -0.007073  0.029534      0.127442  0.004230  ...   
2019-04-30      0.056151  -0.001173  0.020894      0.087364  0.001781  ...   
2019-05-31      0.045780  -0.002382  0.021742      0.075116 -0.001119  ...   
2019-06-30      0.027037   0.008727  0.011859      0.023406  0.002878  ...   

             std_ZTS  min-max_ZTS  mean_mean  mean_std   mean_dd  mean_corr  \
Month                                                                         
2018-01-31  0.010624     0.036959   0.001854  0.003149  0.056860   0.180422   
2018-02-28  0.020803     0.092679  -0.002176  0.003304  0.083620   0.543528   
2018-03-31  0.016602     0.063091  -0.000362  0.002471  0.067023   0.441588   
2018-04-30  0.013100     0.050972   0.000205  0.003118  0.066615   0.398066   
2018-05-31  0.013333     0.059447   0.000689  0.003242  0.063729   0.218755   
2018-06-30  0.010337     0.039441   0.000509  0.002816  0.053663   0.167439   
2018-07-31  0.012461     0.045578   0.001504  0.002903  0.062969   0.130316   
2018-08-31  0.015305     0.089171   0.000811  0.002810  0.055381   0.171057   
2018-09-30  0.007274     0.021116  -0.000016  0.002817  0.048520   0.097006   
2018-10-31  0.017188     0.053390  -0.003131  0.003882  0.090479   0.352152   
2018-11-30  0.016545     0.068998   0.001332  0.003546  0.078153   0.292262   
2018-12-31  0.020937     0.085597  -0.005481  0.003551  0.09595

   Prepared dataset number of rows is: 18
Prepared dataset number of columns is: 1502


In [41]:
def test1(df, col_name):
    """Calculate the Sharpe Ratio difference of MVP and HRP"""
    df.index = pd.to_datetime(df.index, format='%Y/%m/%d')
    df2 = df.groupby(pd.Grouper(freq='M')).mean()
    col_n = 'sharpe_ratio_' + col_name
    df2[col_n] = df.groupby(pd.Grouper(freq='M')).apply(
        lambda x: x.mean()*np.sqrt(12) / x.std()).values
    return df2


metrics = pd.concat([test1(Insample_Result[['MVP']], 'MVP'),
                    test1(Insample_Result[['HRP']], 'HRP')], axis=1)
target = metrics['sharpe_ratio_HRP'] - metrics['sharpe_ratio_MVP']
print("The Difference in Shapre ratio between HRP and MVP.")
display(target)

The Difference in Shapre ratio between HRP and MVP.


Date
2018-01-31    0.600005
2018-02-28    0.296332
2018-03-31   -0.275855
2018-04-30   -0.188617
2018-05-31    0.300619
2018-06-30   -0.199395
2018-07-31    0.540849
2018-08-31    1.239637
2018-09-30    0.841715
2018-10-31   -0.584159
2018-11-30   -0.964496
2018-12-31   -0.131674
2019-01-31    0.617652
2019-02-28   -0.335173
2019-03-31    0.077104
2019-04-30    0.231510
2019-05-31   -0.975094
2019-06-30   -1.060781
Freq: M, dtype: float64

***
<h2 style="color:#007CFE">Explainable AI</h2>

***

To understand why `HRP` **performs better** than `MVP`, we performed `XGBRegressor` on our dataset and applied `SHAP`. Shown below are the plots for explainable AI (XAI).

In [42]:
regressor = XGBRegressor()
regressor.fit(ret3, target)

# Create object that can calculate shap values
shap_explainer = shap.TreeExplainer(regressor)

# Calculate Shap values
shap_values = shap_explainer.shap_values(ret3)

In [43]:
features = ret3.columns
# Comment-out and load the previously saved image to be able to center-align
# shap.summary_plot(shap_values, ret3,
#                   feature_names=features, plot_type="bar")

# Create an HTML img tag to display the image
img_tag = (f'<img src="Global_Shap.png" alt="plots" style="display:block;'
           f'margin-left:auto;margin-right:auto;width:40%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()
fig_caption('Global Feature Importance Plot', '')

In [44]:
# Comment-out and load the previously saved image to be able to center-align
# shap.summary_plot(shap_values, ret3, feature_names=features)

# Create an HTML img tag to display the image
img_tag = (f'<img src="BeeSwarm_Plot_Shap.png" alt="plots" style="display:'
           f'block;margin-left:auto;margin-right:auto;width:40%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()
fig_caption('Bee Swarm Plot', '')

Above is the summary plot of the Shap Value results, highlighting the specific asset (`FLT`, `PXD`, `KSS`) statistic that provides the highest marginal difference to the model. We will examine this top three more closely.

In [45]:
# Comment-out and load the previously saved image to be able to center-align
# shap.dependence_plot(589, shap_values, ret3, feature_names=features)

# Create an HTML img tag to display the image
img_tag = (f'<img src="Dependence_Plot_Shap.png" alt="plots" style="display:'
           f'block;margin-left:auto;margin-right:auto;width:40%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()
fig_caption('Dependence Plot', '')

Examining more the stock price of FleetCor. Show below are the analysis.

<center><img src="./FLT_Analysis.png" width="60%" height="60%"></center>
<center>Image 05. Analysis of FleetCor Technologies, Inc. (FLT) Standard Deviation.</center>

Fleetcor is a US tech company. The movement of this stock highlights the **benefits of diversification** HRP provides. As we can see, MVP overweights the stock about `75%` **higher** than HRP. In times of *high volatility*, the MVP portfolio will bear the weight of that loss, vs. a *dampened effect* on HRP. 

<center><img src="./PXD_Analysis.png" width="60%" height="60%"></center>
<center>Image 06. Analysis of Pioneer Natural Resources Company (PXD) Standard Deviation.</center>

For the second, `SHAP` points toward the stock volatility and `PXD`. However, we see a point toward the MV portfolio in this instance. The mean-variance portfolio allocates practically zero to this particular stock, meaning the *volatility of PXD contributes to the increase* and decrease of the HRP's return. Whereas, for MVP, we can safely ignore its movement. For the HRP scenario, PXD will act as a hedge to another stock in times of volatility.

<center><img src="./KSS_Analysis.png" width="60%" height="60%"></center>
<center>Image 07. Analysis of Kohl's Corporation (KSS) Mean.</center>



Performing EDA on the actual stock price, we can see that the return computations are pretty *stable* while at the same time trending upwards. Because allocation is given to KSS' relatively steady returns, an HRP portfolio reaps the benefit, whereas MVP will not.

<h1 style="background-color:#0531BC; color:#F5F5F1; padding: 20px;">Conclusion</h1>

So let's circle back to our main question, can Hierarchical Risk Parity replace the staple used by Modern Portfolio Theory? The thing is, you will never get a straight answer because the **market will always be dynamic**. However, with the help of `XAI`, we've shown that in our portfolio, **HRP provides a more stable diversification** strategy and covers MVP's shortcomings. While not return target-based, it provides diversification, not at the expense of returns. 

The **mean-variance** strategy is an industry-wide practice because of its **simplicity**, but we believe its also the lack of accessibility of other strategies that drive its prevalence. It used to be that only large firms had access to this technology. But that's not the case at this age. As a result, the playing field is leveled about *knowledge distribution*.

HRP is one among many which may be unboxed now due to `XAI`. From just absolute return percentages, we can examine other metrics to help us make better business decisions. For example, financial products may be improved, and financial literacy encouraged. For clients, it's a new opportunity to understand investment strategies. For institutions, it's a way to tailor fit solutions further.

<h1 style="background-color:#0531BC; color:#F5F5F1; padding: 20px;">Recommendation</h1>

The asset universe utilized in this report was limited to just the `S&P500` member stocks from about two years. As a result, performance comparisons of the Hierarchical Risk Parity portfolio against the Mean-Variance portfolio *have a bias toward the latter* (since equities tend toward returns rather than diversification). Often, a market sell-off sees most stocks moving in the same direction.

The **true benefits of diversification** are only realized when one **moves beyond a single asset class** (in this case, equities); empirically, it is in this scenario that the HRP outperformance can truly be seen. The group recommends the evaluation of a **multi-asset portfolio** constructed with Hierarchical Risk Parity and **subject to portfolio rebalancing**. Assets such as `fixed income`, `commodities`, and `alternative investments` have been known to have statistical properties which differ greatly from equities. Linkages formed in the clustering step of HRP make these differences known immediately based on the class position in the dendrogram. Assessing this portfolio's performance over the long term versus that of a mean-variance portfolio will provide a deeper understanding of the differences between the two strategies and, consequently, clearer factors obtained from applying explainable AI methods.

<h1 style="background-color:#0531BC; color:#F5F5F1; padding: 20px;">References</h1>

[1] Hayes, A. (2023 February 26). *Portfolio Management: Definition, Types, and Strategies.* Investopedia. Retrieved March 16, 2023 from https://www.investopedia.com/terms/p/portfoliomanagement.asp

[2] Capital.com (n.d.). *What is PSE Index (PSEi)?* Retrieved March 17, 2023 from https://capital.com/pse-index-psei-definition

[3] GuidedChoice (n.d.). *Harry Markowitz’s Modern Portfolio Theory: The Efficient Frontier.* Retrieved March 17, 2023 from https://www.guidedchoice.com/video/dr-harry-markowitz-father-of-modern-portfolio-theory/

[4] Jensen, G. (2022, April 09). *Hierarchical Risk Parity on RAPIDS: An ML Approach to Portfolio Allocation.* Retrieved March 17, 2023 from https://developer.nvidia.com/blog/hierarchical-risk-parity-on-rapids-an-ml-approach-to-portfolio-allocation/

[5] Vyas, A. (n.d.). *The Hierarchical Risk Parity Algorithm: An Introduction.* Retrieved March 17, 2023 from https://hudsonthames.org/an-introduction-to-the-hierarchical-risk-parity-algorithm/

[6] López de Prado, M. (2016, May 23). *Building Diversified Portfolios that Outperform Out-of-Sample.* Journal of Portfolio Management, 2016, 42(4), 59–69; https://doi.org/10.3905/jpm.2016.42.4.059. Retrived March 17, 2023 from https://papers.ssrn.com/sol3/papers.cfm?abstract_id=2708678

[7] Fernando, J. (2022, June 06). *Sharpe Ratio Formula and Definition With Examples.* Retrieved March 17, 2023 from https://www.investopedia.com/terms/s/sharperatio.asp

[8] Jaeger, M., Krügel, S., Marinelli, D., Papenbrock, J., Schwendner, P. (2021). Interpretable machine learning for diversified portfolio construction. The Journal of Financial Data Science, 3(3), 31–51. https://doi.org/10.3905/jfds.2021.1.066